packages/clerk-js/src/ui/signUp/utils.ts

```diff
@@ -45,5 +70,75 @@ export function determineFirstPartyFields(environment: EnvironmentResource, hasT
     fields.password = 'required';
   }
 
+  // Enable fields based on current form state
+
+  const shouldShowEmailAddressField =
+    (isEmailAddressUsedForIdentification && hasTicket && hasEmail) ||
+    fields.emailAddress ||
+    (fields.emailOrPhone && activeCommIdentifierType === 'emailAddress');
+
+  if (shouldShowEmailAddressField) {
+    fields.emailAddress ||= 'required'; // Is this accurate?
+  }
+
+  const shouldShowPhoneNumberField =
+    fields.phoneNumber || (fields.emailOrPhone && activeCommIdentifierType === 'phoneNumber');
+
+  if (shouldShowPhoneNumberField) {
+    fields.phoneNumber ||= 'required'; // Is this accurate?
+  }
+
+  // If continuing with an existing sign-up, show only fields absolutely necessary to minimize fiction
+
+  if (signUp) {
+    const hasVerifiedEmail = signUp.verifications?.emailAddress?.status == 'verified';
+    const hasVerifiedPhone = signUp.verifications?.phoneNumber?.status == 'verified';
+    const hasVerifiedExternalAccount = signUp.verifications?.externalAccount?.status == 'verified';
+
+    if (hasVerifiedEmail) {
+      delete fields.emailAddress;
```

Comment:
🔧 Creating the superset of fields and the removing the non-missing and optional ones seems a bit counter intuitive. Isn't it simpler to do the opposite and build the final fields object as the business logic flows?

Reply:
I didn't want to re-write the field determination logic for this feature, so I took a conservative approach.

What I essentially did was:

- Determine fields to show, as before
- If PSU-mode, strip anything that's redundant for completing the sign-up

I would recommend a re-write in a separate PR.

Reply:
The function was rewritten to be more readable, each field is generated in separate methods.